compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/calls/CallableReferenceResolution.kt

```diff
@@ -396,10 +395,67 @@ fun ConeKotlinType.isKCallableType(): Boolean {
 private fun createKPropertyType(
     propertyOrField: FirVariable<*>,
     receiverType: ConeKotlinType?,
-    returnTypeRef: FirResolvedTypeRef
+    returnTypeRef: FirResolvedTypeRef,
+    callInfo: CallInfo,
 ): ConeKotlinType {
     val propertyType = returnTypeRef.type
     return org.jetbrains.kotlin.fir.resolve.createKPropertyType(
-        receiverType, propertyType, isMutable = propertyOrField.isVar
+        receiverType,
+        propertyType,
+        isMutable = propertyOrField.canBeMutableReference(callInfo)
     )
 }
+
+private fun FirVariable<*>.canBeMutableReference(callInfo: CallInfo): Boolean {
+    if (!isVar) return false
+    if (this is FirField) return true
+    if (this is FirProperty && this.hasJvmFieldAnnotation) return true
+    return setter == null || setter!!.isVisible(this, callInfo)
+}
+
+private fun FirPropertyAccessor.isVisible(variable: FirVariable<*>, callInfo: CallInfo): Boolean {
```

Comment:
I wonder if it's possible to reuse the logic from org.jetbrains.kotlin.fir.FirVisibilityChecker

Reply:
Thanks for the pointer! In doing so, though, I found a couple missed implementations.

For `protected` declarations, owner id is a key to determine visibility, and property accessors, including default accessors, don't have `containingClassAttr`. I added that when possible.

For `private` declarations, if they are top-level, their containing file is a key. `FirProvider`, however, didn't have entries (in `callableContainerMap`) for property accessors. I added that too.